scripts/clean_osm_data.py

```diff
@@ -762,31 +785,26 @@ def prepare_generators_df(df_all_generators):
     return df_all_generators
 
 
-def find_first_overlap(geom, country_geoms, default_name):
-    """
-    Return the first index whose shape intersects the geometry.
-    """
-    for c_name, c_geom in country_geoms.items():
-        if not geom.disjoint(c_geom):
-            return c_name
-    return default_name
-
-
 def set_countryname_by_shape(
     df,
     ext_country_shapes,
-    exclude_external=True,
     col_country="country",
 ):
     "Set the country name by the name shape"
-    df[col_country] = [
-        find_first_overlap(
-            row["geometry"],
-            ext_country_shapes,
-            None if exclude_external else row[col_country],
-        )
-        for id, row in df.iterrows()
-    ]
+
+    # use auxiliary spatial join to determine the country of each line, keep first occurrence only
+    df_m = gpd.sjoin_nearest(
+        df.drop(columns=[col_country], errors="ignore"),
+        ext_country_shapes,
+        how="left",
+        max_distance=0.01,
+    ).reset_index()
```

Comment:
Is it possible that no spatial join would be possible for each element of `df`? For example, a distance to the country border will be more than `max_distance` for all OSM-extracted substations? What would happen in this case?

What has been the idea of taking exactly 0.01 as `max_distance`? (If I understand properly, `distance_crs` is used here meaning that it is `EPSG:3857` with meters as unit of measurements, right?)

Reply:
So, when a shape is contained into another, the distance is 0.0.
In the case you mentioned, the joined gdf would not contain all rows.
That however would not throw errors as we only specify the country column only in those indexes that have been matched.

Specifying max distance improve performances. I agree that we should revise the crs, but mainly to avoid the warning as from a functionality perspective any small number should do the job.

This consideration makes me think that we could revise the code to avoid the additional dropping function above by using the max distance wisely, but it would be more difficult to maintain: 1 function with 2 ifs to handle countryname by shape and dropping option.
I'll think about it :)

Reply:
Ahh, ok! I have missed that the distance will be exactly zero when the country shape contains a line or a point. Thanks a lot for the detailed explanation! My feeling is that a comment could be probably clarified a bit (like adding `join with the country shape each spatial element within the shape`).

Agree the a value of `max_distance` is not essential, and its increase would lead only to performance drop. Regarding further improvements, my impression is the approach is non-trivial and elegantly implemented right now :)